### PR TITLE
Feature/9663 Password-Protected-PDF

### PR DESCRIPTION
**Pull request by ArturFriesen**
## Description
Add password entry and error handling for protected PDFs.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-9663

## Screenshots
IMG_7862
IMG_7863

### CODE REVIEW

**Review by NataliaLemmerth — APPROVED** [review]
UA approved

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 1]
```diff
@@ -132,6 +152,17 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 
 	// MARK: - Private
 
+	private func qrCodes(from pdfDocument: PDFDocument) -> [String] {
+		Log.debug("PDF picked will scan for QR codes on all pages", log: .fileScanner)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:156]:
```suggestion
		Log.debug("PDF picked, will scan for QR codes on all pages", log: .fileScanner)
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift [hunk 2]
```diff
@@ -38,6 +38,12 @@ class FileScannerCoordinator {
 			},
 			qrCodesFound: { codes in
 				Log.debug("\(codes.count) codes found", log: .fileScanner)
+			},
+			providePasswordForPDF: { callback in
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift:42]:
```suggestion
			missingPasswordForPDF: { callback in
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 3]
```diff
@@ -15,12 +15,16 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 		showHUD: @escaping () -> Void,
 		hideHUD: @escaping () -> Void,
 		dismiss: @escaping () -> Void,
-		qrCodesFound: @escaping ([String]) -> Void
+		qrCodesFound: @escaping ([String]) -> Void,
+		providePasswordForPDF: @escaping (@escaping (String) -> Void) -> Void,
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:19]:
```suggestion
		missingPasswordForPDF: @escaping (@escaping (String) -> Void) -> Void,
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 4]
```diff
@@ -162,6 +193,8 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 	private let hideHUD: () -> Void
 	private let dismiss: () -> Void
 	private let qrCodesFound: ([String]) -> Void
+	private let providePasswordForPDF: (@escaping (String) -> Void) -> Void
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:196]:
```suggestion
	private let missingPasswordForPDF: (@escaping (String) -> Void) -> Void
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 5]
```diff
@@ -89,18 +91,36 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
+				providePasswordForPDF { [weak self] password in
+					guard let self = self else { return }
+
+					if pdfDocument.unlock(withPassword: password) {
+						Log.debug("PDF successfully unlocked.", log: .fileScanner)
+
+						self.qrCodesFound(self.qrCodes(from: pdfDocument))
+						self.dismiss()
+					} else {
+						Log.debug("PDF unlock failed.", log: .fileScanner)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:108]:
```suggestion
						Log.debug("PDF unlocking failed.", log: .fileScanner)
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 6]
```diff
@@ -15,12 +15,16 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
-		qrCodesFound: @escaping ([String]) -> Void
+		qrCodesFound: @escaping ([String]) -> Void,
+		providePasswordForPDF: @escaping (@escaping (String) -> Void) -> Void,
+		failedToUnlockPDF: @escaping () -> Void
 	) {
 		self.showHUD = showHUD
 		self.hideHUD = hideHUD
 		self.dismiss = dismiss
 		self.qrCodesFound = qrCodesFound
+		self.providePasswordForPDF = providePasswordForPDF
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:26]:
```suggestion
		self.missingPasswordForPDF = missingPasswordForPDF
```

**Review by naveeddotio — APPROVED** [review]
Approved with small comments. Excellent :)

**Review by marcussc — APPROVED** [review]
Very nice :D

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift (hunk 2 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift:42]: good point, renamed it.

**Comment by jucktnich** [comment]
You should disable suggestions for the keyboard I guess.

**Comment by ArturFriesen** [comment]
> You should disable suggestions for the keyboard I guess.

Good point! I added password capabilities to the text field.

**Pull request merged**